backend/module/chaos/serverexperimentation/xds/ecds.go

```diff
@@ -0,0 +1,255 @@
+		if err != nil {
+			logger.Errorw("Invalid http fault config", "config", httpFaultConfig)
+			continue
+		}
+
+		isExternalFault, abort, delay, err = createAbortDelayConfig(httpFaultConfig, downstreamCluster)
+		if err != nil {
+			logger.Errorw("Unable to create ECDS filter fault", "config", httpFaultConfig)
+			continue
+		}
+
+		// We can only perform one fault test per cluster with ECDS. Hence we break.
+		break
+	}
+
+	faultFilter := &gcpFilterFault.HTTPFault{
+		Delay: delay,
+		Abort: abort,
+
+		// override runtimes so that default runtime is not used.
+		DelayPercentRuntime:    delayPercentRuntime,
+		DelayDurationRuntime:   delayDurationRuntime,
+		AbortHttpStatusRuntime: abortHttpStatusRuntime,
+		AbortPercentRuntime:    abortPercentRuntime,
+	}
+
+	var faultFilterName string
+	if isExternalFault {
+		faultFilterName = fmt.Sprintf(FaultFilterConfigNameForExternalFault, upstreamCluster)
+	} else {
+		faultFilterName = FaultFilterConfigNameForInternalFault
+
+		// match downstream cluster with request header else fault will be applied to requests coming from all downstream
+		faultFilter.Headers = []*gcpRoute.HeaderMatcher{
+			{
+				Name: EnvoyDownstreamServiceClusterHeader,
+				HeaderMatchSpecifier: &gcpRoute.HeaderMatcher_ExactMatch{
+					ExactMatch: downstreamCluster,
+				},
+			},
+		}
+	}
+
+	logger.Debugw("ECDS Fault filter", "faultFilter", faultFilter, "upstreamcluster", upstreamCluster)
+
+	serializedFaultFilter, err := proto.Marshal(faultFilter)
+	if err != nil {
+		logger.Warnw("Unable to unmarshal fault filter", "faultFilter", faultFilter)
+		return []gcpTypes.ResourceWithTtl{}
+	}
+
+	return []gcpTypes.ResourceWithTtl{{
+		Resource: &gcpCoreV3.TypedExtensionConfig{
+			Name: faultFilterName,
+			TypedConfig: &any.Any{
+				TypeUrl: faultFilterTypeURL,
+				Value:   serializedFaultFilter,
+			},
+		},
+		Ttl: ttl,
+	}}
+}
+
+func generateEmptyECDSResource(cluster string, ecdsConfig *ECDSConfig, logger *zap.SugaredLogger) []gcpTypes.ResourceWithTtl {
+	faultFilter := &gcpFilterFault.HTTPFault{
+		Delay: DefaultDelayFaultConfig,
+		Abort: DefaultAbortFaultConfig,
+
+		// override runtimes so that default runtime is not used.
+		DelayPercentRuntime:    delayPercentRuntime,
+		DelayDurationRuntime:   delayDurationRuntime,
+		AbortHttpStatusRuntime: abortHttpStatusRuntime,
+		AbortPercentRuntime:    abortPercentRuntime,
+	}
+
+	serializedFaultFilter, err := proto.Marshal(faultFilter)
+	if err != nil {
+		logger.Warnw("Unable to unmarshal fault filter", "faultFilter", faultFilter)
+		return []gcpTypes.ResourceWithTtl{}
+	}
+
+	if _, exists := ecdsConfig.ecdsResourceMap.requestedResourcesMap[cluster]; !exists {
+		logger.Debugw("Cluster not found in ECDS resource map", "cluster", cluster)
+		return []gcpTypes.ResourceWithTtl{}
+	}
+
+	ecdsConfig.ecdsResourceMap.mu.Lock()
+	resourceNames := ecdsConfig.ecdsResourceMap.requestedResourcesMap[cluster]
+	ecdsConfig.ecdsResourceMap.mu.Unlock()
```

Comment:
might be cleaner to expose this as functions on the safe resource map so callers don't have to worry about locking?